main/solution/machine-images/config/infra/files/rstudio/check-idle

```diff
@@ -1,42 +1,53 @@
 #!/usr/bin/env python
 
-
 import os
-import time
+import datetime as dt
+import dateutil.parser as parser
 
 
 LAST_ACTIVE_TIME_FILENAME = '/tmp/last-active-rsession-time'
+ACCESS_LOG_FILENAME = '/var/log/nginx/access.log'
 
-MAX_IDLE_SECONDS = 3600
-
+# To disable auto-stop for RStudio, enter 0 as the value below for MAX_IDLE_MINUTES
+MAX_IDLE_MINUTES = 60
 
-def get_uptime():
-    return float(open('/proc/uptime').read().split()[0])
 
+def get_activity_count():
+    non_ping_count = 0
+    with open(ACCESS_LOG_FILENAME) as access_log:
+        for line in access_log:
+            # nginx sometimes pings the RStudio instance
+            # which does not indicate user activity
+            if "POST /events/get_events" not in line:
+                non_ping_count += 1
+    return (non_ping_count)
 
-def is_rsession(pid):
-    try:
-        cmdline = open(os.path.join('/proc', pid, 'cmdline'), 'rb').read()
-        return 'rsession' in cmdline
-    except IOError:
-        return False
 
+current_time = dt.datetime.now()
+latest_activity_count = get_activity_count()
 
-all_pids = [p for p in os.listdir('/proc') if p.isdigit()]
-rsession_pids = [p for p in all_pids if is_rsession(p)]
+if not os.path.exists(LAST_ACTIVE_TIME_FILENAME):
+    open(LAST_ACTIVE_TIME_FILENAME, 'w').write(
+        str(latest_activity_count) + '\n' + str(current_time))
 
-num_sessions = len(rsession_pids)
+# These will be overwritten by the stored file values
+last_activity_count = 0
+last_active_time = dt.datetime.now()
 
-current_time = time.time()
+with open(LAST_ACTIVE_TIME_FILENAME) as fh:
+    last_activity_count = int(fh.readline())
+    for line in fh:
+        pass
+    last_active_time = parser.parse(line)
 
-if num_sessions or not os.path.exists(LAST_ACTIVE_TIME_FILENAME):
-    open(LAST_ACTIVE_TIME_FILENAME, 'w').write(str(current_time))
-    print('{0} RStudio sessions active at {1}'.format(num_sessions, current_time))
+# Compare with old_activity_count
+# If higher than old count, update file with time and count
+if latest_activity_count > last_activity_count:
```

Comment:
Why is the condition here higher than old count? Not latest_activity_count > 0?

Reply:
If there is any activity on the RStudio platform, the count would increase. Even if the instance is rebooted this count, and the log file from which the count is derived, do not change; the count and logs are simply updated and appended respectively.

Reply:
Ah! Got it. Thanks for the explanation. :)